salt/_modules/metalk8s_volumes.py

```diff
@@ -0,0 +1,360 @@
+        """Path to the backing block device."""
+        return
+
+    @property
+    def is_formatted(self):
+        """Check if the volume is already formatted."""
+        fs_type = self['spec.storageClassName.parameters.fsType']
+        return __salt__['disk.fstype'](self.block_device) == fs_type
+
+    def format(self):
+        """Format the volume.
+
+        The volume is formatted according to its StorageClass.
+        """
+        params = self['spec.storageClassName.parameters']
+        # Options are stored as JSON-encoded list.
+        mkfs_options = json.loads(params['mkfsOptions'])
+        command = ['mkfs']
+        command.extend(['-t', params['fsType']])
+        command.extend(['-U', self['metadata.uid']])
+        command.extend(mkfs_options)
+        command.append(self.block_device)
+        _run_cmd(' '.join(command))
+
+    def __getitem__(self, path):
+        """Return the Volume attribute `attr` from the Volume dict."""
+        return functools.reduce(operator.getitem, path.split('.'), self._volume)
+
+
+# }}}
+# SparseLoopDevice {{{
+
+
+class SparseLoopDevice(Volume):
+    @property
+    def sparse_file(self):
+        return '/var/lib/metalk8s/storage/sparse/{}'.format(
+            self['metadata.name']
+        )
+
+    @property
+    def size(self):
+        return _quantity_to_bytes(self['spec.sparseLoopDevice.size'])
+
+    @property
+    def exists(self):
+        path = self.sparse_file
+        return os.path.isfile(path) and os.path.getsize(path) == self.size
+
+    def create(self):
+        # Try to create a sparse file, don't clobber existing one!
+        try:
+            fd = os.open(self.sparse_file, os.O_WRONLY | os.O_CREAT | os.O_EXCL)
+        except OSError as exn:
+            raise Exception('cannot create sparse file at {}: {}'.format(
+                self.path, exn
+            ))
+        # Set the sparse file capacity.
+        try:
+            os.ftruncate(fd, self.size)
+        except OSError as exn:
+            raise Exception('cannot resize sparse file at {} to {}: {}'.format(
+                self.path, self['spec.sparseLoopDevice.size'], exn
+            ))
+        finally:
+            os.close(fd)
+
+    @property
+    def is_initialized(self):
+        # A sparse loop device is initialized when a sparse file is associated
+        # to a loop device.
+        command = ' '.join(['losetup', '--associated', self.sparse_file])
+        pattern = r'\({}\)'.format(re.escape(self.sparse_file))
+        result  = _run_cmd(command)
+        return re.search(pattern, result['stdout']) is not None
+
+    def initialize(self):
+        # Recent losetup support `--nooverlap` but not the one shipped with
+        # CentOS 7.
+        command = ' '.join(
+            ['losetup', '--find', '--partscan', '--show', self.sparse_file]
+        )
+        result  = _run_cmd(command)
+        return result['stdout'].strip()
+
+    @property
+    def block_device(self):
```

Comment:
I guess this is used to format. I'd rather not do it that way, but format *what we know* (i.e., what's passed in by a user with a `rawBlockDevice`, or in this case, the backing sparse file, which can be formatted in-place, rather than formatting after creating the block device, which is more failure-prone).

Reply:
Done.